### DISCUSSION TITLE
Road to v1

### DISCUSSION THREAD

**Discussion by timostamm** [connectrpc/connect-es#315] [Announcements]
We don't have any outstanding features blocking a v1 of Connect for Web, but there is outstanding work for Protobuf-ES and Connect for Node.js. We plan to ship a v1 of all three around the same time very early next year. We'll keep this issue updated as we make progress. 

### Connect for Node.js

There will be a new package `@bufbuild/connect-node`, which will export client transports for the Connect, gRPC, and gRPC-web protocols, using the Node.js core packages `http`, `https`, and `http2` instead of the fetch API. 

The package will also enable handlers for _serving_ Connect, gRPC, and gRPC-web, so that they can easily be plugged into a vanilla Node.js server, an express or Next.js server.

Because this will enable client-streaming and bidi-streaming RPCs, we are making changes to the `Transport` and `Interceptor` types to support them. The new shared types live in `@bufbuild/connect-core`. Once they are stabilized, we will cut a release of`@bufbuild/connect-web` that uses them.

Update: [@bufbuild/connect-node](https://www.npmjs.com/package/@bufbuild/connect-node) is now available as a preview. 

### Protobuf-ES - reached [v1](https://github.com/bufbuild/protobuf-es/releases/tag/v1.0.0)

If you use Connect for Web, you are using the generated messages from Protobuf-ES. In your UI layer, you might prefer working with plain objects over working with classes, so we are currently investigating generating interfaces in addition to the generated classes. This is tracked in [bufbuild/protobuf-es#230](https://github.com/bufbuild/protobuf-es/issues/230).

Note that we are also considering other features like [ECMAScript Map for protobuf map fields](https://github.com/bufbuild/protobuf-es/issues/258) and support for [proto2 extensions](https://github.com/bufbuild/protobuf-es/issues/86), but we will most likely address those post v1.

### Connect for Web

If you are using `@bufbuild/connect-web`, only interceptors will change slightly (we will provide a guide for switching over). Everything else will continue to work as is, without any changes to your code.

**Comment by timostamm** [comment]
The past six weeks have been quite busy. We're glad to share that we have clients and servers for the Connect, gRPC, and the gRPC-web protocol roughly working in Node.js :tada:

A couple of features are still missing, for example compression support in all protocols, and support for deadlines. After implementing those missing features, we will focus on streamlining the implementation and shaking out bugs. We plan to publish a first beta in January. We're thrilled about the amount of feedback we've received already. We truly appreciate that, thank you!

Regarding Protobuf-ES, we are about to release a v1 in the next couple of days. That doesn't mean we are done, but we think it is mature enough to cut a stable version.

**Reply by fubhy** [reply]
Right on time for Christmas! Wonderful news Timo! Kudos to you and your team <3

**Reply by nicole0707** [reply]
Just wondering when `@bufbuild/connect-node` will be released, cheers.

**Reply by timostamm** [reply]
Hey Nicole, the functionality for all three protocols (Connect, gRPC, gRPC-web) is there. But as usual, the details take time. We're still looking to release a first beta in January, but it's most likely going to be the latter halve.

**Reply by ghost** [reply]
Hello! First of all, thank you very much for this work. :heart: 
I looked at the `connect-node` code out of curiosity, and I couldn't figure out if and how cancellation is handled. Will it eventually use `AbortSignal`?

**Reply by timostamm** [reply]
Hey Johyn, that's right, all methods on clients take an optional `AbortSignal`. You can find details in the [docs](https://connect.build/docs/web/errors#cancellation) - the same will apply to Node clients.

On the server, you can always throw a ConnectError, for example with `Code.Aborted`. We will also support [deadlines](https://grpc.io/blog/deadlines/), but we might cut a first beta without, and add it as a follow-up 🙂

**Reply by ghost** [reply]
Hey! Thank you very much for this information. But there is still one point I don't understand, how is the server informed when a call is cancelled? Thank you very much in advance.

**Reply by timostamm** [reply]
Cancelling a request via `AbortSignal` is somewhat similar from turning off power to your computer. No "cancel" message is sent to the server. There is nothing special to it, it is just plain HTTP.

**Reply by ghost** [reply]
OK thank you very much, I will go and study the gRPC protocol in detail.
Using `connect-node`, how does the server, for example in the case of a server streaming call, know when the stream is closed? Let's say I need to create and then release a resource, how do I know when to release this resource?
Thanks a lot in advance. :pray:

**Comment by matamatanot** [comment]
Does connect-web (or connect-nodejs or maybe connect-edge) aim to be WinterCG spec compliant?
I am a Next.js user and would like to run APIRoutes in an Edge environment and use the connect protocol from that environment.

[proposal-common-minimum-api](https://github.com/wintercg/proposal-common-minimum-api) is here.

**Reply by shinebayar-g** [reply]
This would be game changer. [Found this](https://community.cloudflare.com/t/can-i-make-a-grpc-request-from-a-worker/157450/5). They say `grpc-web` should work.

**Reply by timostamm** [reply]
Hey @matamatanot, that's an excellent question. Everything we've done so fair aims to be as universal as possible. 

We took a stab at using WHATWG streams ([PR](https://github.com/bufbuild/connect-web/pull/408)), but that actually showed a couple of problems in practice (described [here](https://github.com/bufbuild/connect-web/pull/422)).

We have since switched handlers to asynchronous iterables ([PR](https://github.com/bufbuild/connect-web/pull/430)), and are currently doing the same for clients. Because asynchronous iterables are a (widely supported) language feature, it is relatively straight-forward take an arbitrary stream implementation like WHATWG streams or Node.js streams, and make the stream available as an asynchronous iterable. The Node.js specific code is actually a rather slim layer that should be straight-forward to be adapted to - let's say the Deno standard lib. 

So yes, we aim to be compliant to the Minimum Common Web Platform API. But please be aware that the connect-node package _has_ to be specific to Node.js to provide a decent developer experience, and can't run on an arbitrary platform without specific bindings. We're looking forward to get this out the door and get feedback from the community, including learning where you would like to use it besides Node.js.

**Reply by shinebayar-g** [reply]
That makes sense, thanks for the update @timostamm .

**Comment by timostamm** [comment]
A quick update: 

We had to go through a couple of iterations of a streams abstraction, but the dust is settling 😄 
A first version of `@bufbuild/connect-node` will be available next week.

**Comment by ghost** [comment]
Hello, thank you very much for this work!

I don't know if this is the best place for this, but I thought I'd share my use-case with `connect-node`, as I'm not sure it's currently feasible.

I actually have a server stream RPC named `listenToChannel` which is intended to listen to a sub pub topic. The current implementation, using `grpc-js` and a wrapper to support async iterables, looks like this:

```ts
const service: PubSubService {
  async* listenToChannel(request, { signal }) {
    const iterable = pubSubClient.topic(request.topic, { signal });
    
    for (const message of iterable) {
      yield new ListenToChannelResponse({ message });
    }
  },
};
```

The idea is that every request uses the pub sub client to start receiving messages on a topic. The signal, which is an `AbortSignal`, is passed to the pub sub client so that the pub sub client can be told to stop listening and release what needs to be released once the grpc stream is closed.

Will this use case be implementable using `connect-node`? This would be the only point that would prevent us from migrating. If you are interested, I can try to make a PR to implement this feature.

Thank you very much in advance!

**Reply by timostamm** [reply]
Johyn, we do not support this use case yet, but we intend to. In any streaming RPC, you definitely want good control over the lifetime and cancellation! 

gRPC and Connect have the concept of [deadlines](https://grpc.io/blog/deadlines/), but there is also the potential for the client to just hang up. In those situations, we need to ensure that the RPC implementation is notified, and has the ability to clean up, for example close a database connection, an open file, or - in your case - a downstream subscription.

All handlers already take a second argument, a `HandlerContext` that already gives you access to headers and trailers. The plan is to add an AbortSignal that you can pass downstream, to ensure that long-running tasks can be notified that the RPC is dead. 

But in many cases, there isn't much you can do with an `AbortSignal` except calling `signal.throwIfAborted()` in your implementation. This doesn't seem ideal, so we have been looking into `AsyncIterator.throw`, to automatically cancel RPC implementations on `yield`. In many cases, that should be sufficient on it's own.

**Reply by Sanger2000** [reply]
Any updates on this? What is the best way at the moment to release resources if a client has closed the connection (mainly via `AbortSignal`)

**Reply by timostamm** [reply]
Hey @Sanger2000, the feature was merged in https://github.com/bufbuild/connect-es/pull/591 last week, but has not been released yet. The link has a more detailed explanation and examples.

**Reply by ghost** [reply]
Hello @timostamm! Reading the code, I have the impression that this signal will not be aborted when the request is closed, only when the deadline is triggered. Am I wrong?

**Reply by smaye81** [reply]
Hi @johynpapin @Sanger2000. Hardening our clients and handlers is something we are working on currently. There are a few scenarios we want to address and write tests for prior to a v1 release and this is on that roadmap. You are correct that this PR involves deadlines and not explicit cancels, so it may not handle your exact use case. However, we're hoping to account for explicit cancels in some upcoming improvements.

**Comment by timostamm** [comment]
We just release the Preview of Connect for Node.js in [v0.7.0](https://github.com/bufbuild/connect-web/releases/tag/v0.7.0) 🎉

You can find a simple example [here](https://github.com/bufbuild/connect-web/tree/main/packages/example), and some documentation will be added to https://connect.build/ shortly.

**Comment by timostamm** [comment]
Connect for Node.js is now officially released 🎉 

> Connect is now full-stack TypeScript! We're excited to launch the beta of Connect for Node.js today, a library for serving Connect, gRPC, and gRPC-Web APIs using Node.js: https://buf.build/blog/connect-node-beta

**Comment by baldr9** [comment]
Does buf Connect support Flutter?

There is a Flutter package that references grpc-web support: https://pub.dev/packages/grpc 
There is a buf connect to crosstest the grpc-Web: https://github.com/bufbuild/connect-crosstest

**Reply by timostamm** [reply]
There is no implementation for Flutter specifically yet. As of today, we have implementations for:
- [Go](https://github.com/bufbuild/connect-go) - clients and servers that speak all three protocols: Connect, gRPC, gRPC-web
- This repository - clients and server for all three protocols (but note that browser clients can only do Connect and gRPC-web)
- [Kotlin](https://github.com/bufbuild/connect-kotlin) - clients for all three protocols
- [Swift](https://github.com/bufbuild/connect-swift) - clients for gRPC-web and Connect

That means you can spin up a connect-go or connect-es server and talk to it from any gRPC, gRPC-web, or Connect client. This includes any of our clients for Kotlin, Swift, Go, web browsers, or Node.js, but also any other gRPC or gRPC-web client like the grpc Dart package.

**Comment by timostamm** [comment]
As you probably noticed, we took a bit of a pivot with [v0.8.0](https://github.com/bufbuild/connect-es/releases/tag/v0.8.0). The `Handler`s we were creating for vanilla Node.js didn't work out nicely with common middleware like `body-parser`, or with frameworks like Fastify, which wrap the built-in request/response types.

To solve this problem, we introduced `ConnectRouter`, so that we have just a single API for users. Server plugins take the role of adapting the routes for various frameworks, giving us more flexibility to support their idiosyncrasies.

The next steps for us are:
- provide tooling and documentation for testing
- support SSR
- improve stability (we have some open issues about unhandledRejection and stream connections hanging, and there's an issue with Node.js v19)

We consider all three to be required for a v1.

We are also currently investigating support for other runtimes, but can't tell whether it will be part of a v1 or a follow-up.

**Comment by timostamm** [comment]
Hey there, a quick update: We are still crunching on the bullet points listed above:
- provide tooling and documentation for testing  
  We have [added an in-memory Transport](https://github.com/bufbuild/connect-es/pull/548) that helps testing an application without hitting the network. It is documented it [here](https://connect.build/docs/node/testing).
- support SSR  
  As a first step, the latest release [v0.9.0](https://github.com/bufbuild/connect-es/releases/tag/v0.9.0) includes an [option on the Transport to provide your own `fetch`](https://github.com/bufbuild/connect-es/pull/618).  
- improve stability  
  The latest release [v0.9.0](https://github.com/bufbuild/connect-es/releases/tag/v0.9.0) contains a lot of fixes for better stability and fully supports timeouts, a.k.a [deadlines in gRPC](https://grpc.io/blog/deadlines/). We are considering to add [keepalive with HTTP/2 ping frames](https://github.com/bufbuild/connect-es/issues/627) for parity with gRPC.

Also note that the Connect protocol now supports GET requests in connect-es. You can find out more in [the documentation](https://connect.build/docs/node/get-requests-and-caching)).

**Comment by timostamm** [comment]
Hey there, we just cut the v1 release yesterday 🎉 

For testing, we document the tooling and options for web applications [here](https://connectrpc.com/docs/web/testing), and for Node.js client and servers [here](https://connectrpc.com/docs/node/testing). Note that we also have an integration with Playwright in [@connectrpc/connect-playwright](https://www.npmjs.com/package/@connectrpc/connect-playwright), and we plan to [support MSW](https://github.com/connectrpc/connect-es/issues/825) as well.

For SSR, most frameworks come with their bespoke technique, although RSC will potentially change the landscape. Connect-ES provides some building blocks for use with SSR boundaries. We provide documentation and examples [here](https://connectrpc.com/docs/web/ssr).

**Reply by nicole0707** [reply]
In terms of RPC testing, I reckon `createRouterTransport` would provide more benefit than `msw`, like reuse interceptors. Just curious, which case is recommended to use  `msw`?

**Reply by timostamm** [reply]
Service Workers can intercept any request a browser makes, just by registering the worker. Requests can be intercepted conditionally - for example you can intercept requests with a specific payload.

So if you want to run tests against a real backend, but intercept only select requests, it gives you some options that are otherwise a bit tricky to achieve. 

For mocking RPC, `createRouterTransport` is the more straight-forward solution.
